src/lib/operators/aggregate_hash.cpp

```diff
@@ -647,12 +698,13 @@ std::shared_ptr<const Table> AggregateHash::_on_execute() {
       break;
     default:
       PerformanceWarning("No std::array implementation initialized - falling back to vector");
-      _aggregate<std::vector<AggregateKeyEntry>>();
+      _aggregate<boost::container::small_vector<AggregateKeyEntry, 4>>();
```

Comment:
(5) Is small_vector faster than vector? Is `array<..., 2>` still useful?

Reply:
Is small_vector faster than vector? No!

https://github.com/hyrise/hyrise/compare/61bd7a4..6c65003

```
+----------++----------+---------+--------++----------+----------+--------+---------+
| Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
|          ||      old |     new |        ||      old |      new |        |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
| TPC-H 01 ||   1651.5 |  1891.6 |  +15%  ||     0.61 |     0.53 |  -13%  |  0.0000 |
| TPC-H 02 ||     11.8 |    14.0 |  +18%  ||    84.50 |    71.49 |  -15%  |  0.0000 |
| TPC-H 03 ||    178.4 |   176.7 |   -1%  ||     5.61 |     5.66 |   +1%  |  0.0054 |
| TPC-H 04 ||    152.0 |   163.7 |   +8%  ||     6.58 |     6.11 |   -7%  |  0.0000 |
| TPC-H 05 ||    470.9 |   500.5 |   +6%  ||     2.12 |     2.00 |   -6%  |  0.0000 |
| TPC-H 06 ||      7.9 |     8.6 |   +9%  ||   126.54 |   115.78 |   -8%  |  0.0000 |
| TPC-H 07 ||    126.2 |   134.3 |   +6%  ||     7.93 |     7.45 |   -6%  |  0.0000 |
| TPC-H 08 ||    156.3 |   164.7 |   +5%  ||     6.40 |     6.07 |   -5%  |  0.0000 |
| TPC-H 09 ||   1013.1 |  1051.7 |   +4%  ||     0.99 |     0.95 |   -4%  |  0.0000 |
| TPC-H 10 ||    511.9 |   539.4 |   +5%  ||     1.95 |     1.85 |   -5%  |  0.0000 |
| TPC-H 11 ||     16.5 |    18.5 |  +12%  ||    60.71 |    54.11 |  -11%  |  0.0000 |
| TPC-H 12 ||     90.0 |    95.7 |   +6%  ||    11.11 |    10.44 |   -6%  |  0.0000 |
| TPC-H 13 ||    705.9 |   738.3 |   +5%  ||     1.42 |     1.35 |   -4%  |  0.0000 |
| TPC-H 14 ||     38.8 |    40.6 |   +5%  ||    25.77 |    24.64 |   -4%  |  0.0000 |
| TPC-H 15 ||     17.6 |    18.8 |   +7%  ||    56.89 |    53.03 |   -7%  |  0.0000 |
| TPC-H 16 ||    178.3 |   125.1 |  -30%  ||     5.61 |     8.00 |  +43%  |  0.0000 |
| TPC-H 17 ||     54.4 |    54.6 |   +0%  ||    18.39 |    18.31 |   -0%  |  0.0724 |
| TPC-H 18 ||   1312.0 |  1338.8 |   +2%  ||     0.76 |     0.75 |   -2%  |  0.0010 |
| TPC-H 19 ||     75.3 |    81.4 |   +8%  ||    13.28 |    12.29 |   -7%  |  0.0000 |
| TPC-H 20 ||     31.7 |    34.7 |   +9%  ||    31.54 |    28.83 |   -9%  |  0.0000 |
| TPC-H 21 ||    822.2 |   868.5 |   +6%  ||     1.22 |     1.15 |   -5%  |  0.0000 |
| TPC-H 22 ||    126.3 |   131.1 |   +4%  ||     7.92 |     7.63 |   -4%  |  0.0000 |
+----------++----------+---------+--------++----------+----------+--------+---------+
| Sum      ||   7748.8 |  8191.3 |   +6%  ||          |          |        |         |
| Geomean  ||          |         |        ||          |          |   -4%  |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
```

However, if we use small_vector and drop array<..., 2> it's slightly beneficial.

https://github.com/hyrise/hyrise/compare/61bd7a4..f6158a4

```
+----------++----------+---------+--------++----------+----------+--------+---------+
| Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
|          ||      old |     new |        ||      old |      new |        |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
| TPC-H 01 ||   1651.5 |  1815.0 |  +10%  ||     0.61 |     0.55 |   -9%  |  0.0000 |
| TPC-H 02 ||     11.8 |    11.6 |   -2%  ||    84.50 |    86.34 |   +2%  |  0.0766 |
| TPC-H 03 ||    178.4 |   164.9 |   -8%  ||     5.61 |     6.06 |   +8%  |  0.0000 |
| TPC-H 04 ||    152.0 |   152.1 |   +0%  ||     6.58 |     6.57 |   -0%  |  0.4789 |
| TPC-H 05 ||    470.9 |   475.9 |   +1%  ||     2.12 |     2.10 |   -1%  |  0.0126 |
| TPC-H 06 ||      7.9 |     7.8 |   -2%  ||   126.54 |   128.69 |   +2%  |  0.0000 |
| TPC-H 07 ||    126.2 |   124.3 |   -1%  ||     7.93 |     8.04 |   +2%  |  0.1594 |
| TPC-H 08 ||    156.3 |   153.1 |   -2%  ||     6.40 |     6.53 |   +2%  |  0.0000 |
| TPC-H 09 ||   1013.1 |  1024.1 |   +1%  ||     0.99 |     0.98 |   -1%  |  0.0049 |
| TPC-H 10 ||    511.9 |   316.8 |  -38%  ||     1.95 |     3.16 |  +62%  |  0.0000 |
| TPC-H 11 ||     16.5 |    16.5 |   -0%  ||    60.71 |    60.73 |   +0%  |  0.8249 |
| TPC-H 12 ||     90.0 |    90.7 |   +1%  ||    11.11 |    11.03 |   -1%  |  0.0001 |
| TPC-H 13 ||    705.9 |   707.1 |   +0%  ||     1.42 |     1.41 |   -0%  |  0.2351 |
| TPC-H 14 ||     38.8 |    38.0 |   -2%  ||    25.77 |    26.32 |   +2%  |  0.0000 |
| TPC-H 15 ||     17.6 |    17.8 |   +2%  ||    56.89 |    56.01 |   -2%  |  0.0000 |
| TPC-H 16 ||    178.3 |   119.2 |  -33%  ||     5.61 |     8.39 |  +50%  |  0.0000 |
| TPC-H 17 ||     54.4 |    54.1 |   -1%  ||    18.39 |    18.49 |   +1%  |  0.0130 |
| TPC-H 18 ||   1312.0 |  1331.0 |   +1%  ||     0.76 |     0.75 |   -1%  |  0.0049 |
| TPC-H 19 ||     75.3 |    74.8 |   -1%  ||    13.28 |    13.37 |   +1%  |  0.0000 |
| TPC-H 20 ||     31.7 |    24.8 |  -22%  ||    31.54 |    40.38 |  +28%  |  0.0000 |
| TPC-H 21 ||    822.2 |   818.9 |   -0%  ||     1.22 |     1.22 |   +0%  |  0.5410 |
| TPC-H 22 ||    126.3 |   124.6 |   -1%  ||     7.92 |     8.03 |   +1%  |  0.0000 |
+----------++----------+---------+--------++----------+----------+--------+---------+
| Sum      ||   7748.8 |  7663.0 |   -1%  ||          |          |        |         |
| Geomean  ||          |         |        ||          |          |   +5%  |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
```

Reply:
Mmh, the last one is interesting. In that case, we would only have specializations for 0, 1, and 2+ group by columns, right?

Reply:
seems that in most of the cases, when there are more than 2 group by columns, there are actually more than 4.
this would explain why the small_vector makes it slower in the first place.

initially, dropping array<...,2> was an option in case that the small_vector improves performance. however, since the small_vector seems to be slower, i don't fully understand why dropping the array specialization can be faster.

Reply:
Nemea:

```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | TPC-H 01 ||    821.3 |   828.5 |   +1%  ||     1.22 |     1.21 |   -1%  |  0.1939 |
 | TPC-H 02 ||      4.6 |     4.5 |   -1%  ||   218.39 |   220.06 |   +1%  |  0.3221 |
 | TPC-H 03 ||     81.4 |    80.6 |   -1%  ||    12.28 |    12.41 |   +1%  |  0.0000 |
 | TPC-H 04 ||     65.9 |    65.4 |   -1%  ||    15.18 |    15.28 |   +1%  |  0.0000 |
+| TPC-H 05 ||    223.3 |   212.5 |   -5%  ||     4.48 |     4.71 |   +5%  |  0.0000 |
 | TPC-H 06 ||      3.2 |     3.2 |   -1%  ||   310.77 |   312.87 |   +1%  |  0.0000 |
 | TPC-H 07 ||     57.0 |    56.9 |   -0%  ||    17.54 |    17.59 |   +0%  |  0.7858 |
 | TPC-H 08 ||     71.7 |    72.4 |   +1%  ||    13.95 |    13.81 |   -1%  |  0.0376 |
 | TPC-H 09 ||    508.6 |   505.5 |   -1%  ||     1.97 |     1.98 |   +1%  |  0.0727 |
 | TPC-H 10 ||    249.0 |   248.7 |   -0%  ||     4.02 |     4.02 |   +0%  |  0.9052 |
 | TPC-H 11 ||      8.3 |     8.2 |   -1%  ||   120.62 |   122.17 |   +1%  |  0.0000 |
 | TPC-H 12 ||     42.4 |    42.2 |   -0%  ||    23.60 |    23.72 |   +0%  |  0.0000 |
 | TPC-H 13 ||    354.9 |   352.0 |   -1%  ||     2.82 |     2.84 |   +1%  |  0.0000 |
 | TPC-H 14 ||     21.2 |    21.2 |   -0%  ||    47.06 |    47.27 |   +0%  |  0.0001 |
 | TPC-H 15 ||      7.7 |     7.7 |   -1%  ||   129.04 |   130.57 |   +1%  |  0.0000 |
 | TPC-H 16 ||     91.2 |    90.4 |   -1%  ||    10.96 |    11.06 |   +1%  |  0.0005 |
 | TPC-H 17 ||     16.4 |    16.4 |   -0%  ||    60.87 |    61.08 |   +0%  |  0.0725 |
+| TPC-H 18 ||    731.0 |   676.9 |   -7%  ||     1.37 |     1.48 |   +8%  |  0.0000 |
 | TPC-H 19 ||     27.8 |    27.7 |   -0%  ||    35.95 |    36.05 |   +0%  |  0.0000 |
 | TPC-H 20 ||     15.7 |    15.6 |   -0%  ||    63.80 |    64.12 |   +1%  |  0.0536 |
 | TPC-H 21 ||    414.8 |   410.8 |   -1%  ||     2.41 |     2.43 |   +1%  |  0.0436 |
 | TPC-H 22 ||     50.0 |    49.4 |   -1%  ||    20.01 |    20.23 |   +1%  |  0.0000 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Sum      ||   3867.4 |  3796.6 |   -2%  ||          |          |        |         |
 | Geomean  ||          |         |        ||          |          |   +1%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```

Again, not as big of an improvement, but if we can remove that code path without any negative performance impact, it reduces the compile time, which is a good thing.

Reply:
Confirmed on nemea:

```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
+| TPC-H 01 ||    810.9 |   741.5 |   -9%  ||     1.23 |     1.35 |   +9%  |  0.0000 |
+| TPC-H 02 ||      5.2 |     4.9 |   -5%  ||   193.44 |   203.85 |   +5%  |  0.0000 |
+| TPC-H 03 ||     83.8 |    75.6 |  -10%  ||    11.93 |    13.22 |  +11%  |  0.0000 |
 | TPC-H 04 ||     68.4 |    68.5 |   +0%  ||    14.62 |    14.60 |   -0%  |  0.0933 |
 | TPC-H 05 ||    232.3 |   230.8 |   -1%  ||     4.30 |     4.33 |   +1%  |  0.0830 |
+| TPC-H 06 ||      3.5 |     3.2 |  -10%  ||   282.34 |   313.62 |  +11%  |  0.0000 |
 | TPC-H 07 ||     58.5 |    58.1 |   -1%  ||    17.08 |    17.21 |   +1%  |  0.4060 |
 | TPC-H 08 ||     76.5 |    76.5 |   -0%  ||    13.07 |    13.08 |   +0%  |  0.9171 |
 | TPC-H 09 ||    517.0 |   519.1 |   +0%  ||     1.93 |     1.93 |   -0%  |  0.3035 |
+| TPC-H 10 ||    262.4 |   164.5 |  -37%  ||     3.81 |     6.08 |  +59%  |  0.0000 |
 | TPC-H 11 ||      8.9 |     8.7 |   -3%  ||   112.24 |   115.48 |   +3%  |  0.0000 |
 | TPC-H 12 ||     43.6 |    43.6 |   +0%  ||    22.96 |    22.95 |   -0%  |  0.8885 |
 | TPC-H 13 ||    351.5 |   352.7 |   +0%  ||     2.85 |     2.84 |   -0%  |  0.0549 |
 | TPC-H 14 ||     21.1 |    21.1 |   -0%  ||    47.39 |    47.46 |   +0%  |  0.2387 |
 | TPC-H 15 ||      7.7 |     7.7 |   +1%  ||   130.28 |   129.08 |   -1%  |  0.0000 |
+| TPC-H 16 ||     93.8 |    63.6 |  -32%  ||    10.66 |    15.72 |  +47%  |  0.0000 |
 | TPC-H 17 ||     19.5 |    19.2 |   -1%  ||    51.38 |    52.13 |   +1%  |  0.0000 |
 | TPC-H 18 ||    697.6 |   671.8 |   -4%  ||     1.43 |     1.49 |   +4%  |  0.0003 |
 | TPC-H 19 ||     31.4 |    31.4 |   +0%  ||    31.89 |    31.86 |   -0%  |  0.0799 |
+| TPC-H 20 ||     16.5 |    12.6 |  -24%  ||    60.63 |    79.63 |  +31%  |  0.0000 |
 | TPC-H 21 ||    420.9 |   426.4 |   +1%  ||     2.38 |     2.35 |   -1%  |  0.0081 |
 | TPC-H 22 ||     50.5 |    49.9 |   -1%  ||    19.79 |    20.04 |   +1%  |  0.0000 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
+| Sum      ||   3881.3 |  3651.3 |   -6%  ||          |          |        |         |
+| Geomean  ||          |         |        ||          |          |   +7%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```